File: server-csharp/Magician/Methods.cs
using System.Numerics;
using SpacetimeDB;

public static partial class Module
{
    public static void AdjustGrounded(ReducerContext ctx, DbVector3 MoveVelocity, ref Magician Magician)
    {

        if (Magician.KinematicInformation.Grounded is true) {
             Magician.KinematicInformation.Falling = false;
            RemoveSubscriber(GetPermissionEntry(Magician.PlayerPermissionConfig, "CanJump").Subscribers, "Jump");
            RemoveSubscriber(GetPermissionEntry(Magician.PlayerPermissionConfig, "CanCrouch").Subscribers, "Jump"); 
        }
        else {
            Magician.KinematicInformation.Falling = MoveVelocity.y < -2f;
            AddSubscriberUnique(GetPermissionEntry(Magician.PlayerPermissionConfig, "CanJump").Subscribers, "Jump");
            AddSubscriberUnique(GetPermissionEntry(Magician.PlayerPermissionConfig, "CanCrouch").Subscribers, "Jump");   
        }

    }

    public static void ResolveContacts(ref Magician CharacterLocal, List<CollisionContact> Contacts, DbVector3 InputVelocity)
    {
        DbVector3 WorldUp = new(0f, 1f, 0f);

        float MinGroundDot = 0.75f;

        float DepthEpsilon = 2e-3f;
        float MaxDepth = 0.08f;

        float CorrectionFactor = 0.5f;
        float TargetPenetration = 0.01f;

        float MaxPositionCorrection = 0.015f;

        float GroundStickUpThreshold = 0.03f;
        float InputUpCancelThreshold = 0.03f;

        DbVector3 CorrectedVelocity = InputVelocity;

        bool IsGroundedOnMap = false;
        DbVector3 GroundSupportNormal = WorldUp;

        bool HasAnyPositionCorrection = false;
        DbVector3 TotalPositionCorrection = new(0f, 0f, 0f);

        foreach (CollisionContact Contact in Contacts)
        {
            DbVector3 Normal = Contact.Normal;

            float UpDot = Dot(Normal, WorldUp);
            bool IsWalkableMapContact = Contact.CollisionType == CollisionEntryType.Map && UpDot >= MinGroundDot;

            if (IsWalkableMapContact)
            {
                IsGroundedOnMap = true;
                GroundSupportNormal = Normal;
            }

            float NormalVelocityComponent = Dot(Normal, CorrectedVelocity);
            if (NormalVelocityComponent < 0f)
                CorrectedVelocity = Sub(CorrectedVelocity, Mul(Normal, NormalVelocityComponent));

            float Depth = Contact.PenetrationDepth - TargetPenetration;
            if (Depth > DepthEpsilon)
            {
                if (Depth > MaxDepth) Depth = MaxDepth;
                HasAnyPositionCorrection = true;
                TotalPositionCorrection = Add(TotalPositionCorrection, Mul(Normal, Depth));
            }
        }

        if (HasAnyPositionCorrection)
        {
            float CorrectionMagnitudeSq = Dot(TotalPositionCorrection, TotalPositionCorrection);
            if (CorrectionMagnitudeSq > 1e-8f)
            {
                float CorrectionMagnitude = MathF.Sqrt(CorrectionMagnitudeSq);
                if (CorrectionMagnitude > MaxPositionCorrection)
                    TotalPositionCorrection = Mul(Normalize(TotalPositionCorrection), MaxPositionCorrection);

                TotalPositionCorrection = Mul(TotalPositionCorrection, CorrectionFactor);
                CharacterLocal.Position = Add(CharacterLocal.Position, TotalPositionCorrection);
            }
        }

        if (IsGroundedOnMap)
        {
            float DesiredHorizontalSpeedSq = InputVelocity.x * InputVelocity.x + InputVelocity.z * InputVelocity.z;

            if (DesiredHorizontalSpeedSq < 0.001f)
            {
                CorrectedVelocity.x = 0f;
                CorrectedVelocity.z = 0f;
            }
            else
            {
                float DesiredHorizontalSpeed = MathF.Sqrt(DesiredHorizontalSpeedSq);
                float PreservedCorrectedY = CorrectedVelocity.y;

                DbVector3 TangentVelocity = Sub(CorrectedVelocity, Mul(GroundSupportNormal, Dot(CorrectedVelocity, GroundSupportNormal)));
                float TangentSpeedSq = Dot(TangentVelocity, TangentVelocity);

                if (TangentSpeedSq > 1e-8f)
                {
                    DbVector3 TangentDirection = Mul(TangentVelocity, 1f / MathF.Sqrt(TangentSpeedSq));

                    float TangentDirectionHorizontalSq = TangentDirection.x * TangentDirection.x + TangentDirection.z * TangentDirection.z;
                    if (TangentDirectionHorizontalSq > 1e-8f)
                    {
                        float TangentDirectionHorizontal = MathF.Sqrt(TangentDirectionHorizontalSq);
                        float Scale = DesiredHorizontalSpeed / TangentDirectionHorizontal;

                        DbVector3 DesiredTangentVelocity = Mul(TangentDirection, Scale);

                        CorrectedVelocity.x = DesiredTangentVelocity.x;
                        CorrectedVelocity.z = DesiredTangentVelocity.z;
                        CorrectedVelocity.y = PreservedCorrectedY;
                    }
                }
            }

            if (CorrectedVelocity.y <= GroundStickUpThreshold)
                CorrectedVelocity.y = 0f;

            if (InputVelocity.y <= InputUpCancelThreshold)
                CharacterLocal.Velocity.y = 0f;
        }

        CharacterLocal.IsColliding = Contacts.Count > 0;
        CharacterLocal.CorrectedVelocity = CorrectedVelocity;
        CharacterLocal.KinematicInformation.Grounded = CharacterLocal.KinematicInformation.Grounded || IsGroundedOnMap;
    }
    private static bool TryBuildContactForEntry(ReducerContext Ctx, ref Magician CharacterLocal, CollisionEntry CollisionEntry, List<CollisionContact> Contacts)
    {
        DbVector3 PositionA = CharacterLocal.Position;
        float YawRadiansA = ToRadians(CharacterLocal.Rotation.Yaw);

        if (CollisionEntry.Type == CollisionEntryType.Magician)
        {
            Magician OtherMagician = Ctx.Db.magician.Id.Find(CollisionEntry.Id) ?? throw new Exception("Colliding Magician Not Found");
            if (OtherMagician.Id == CharacterLocal.Id) return false;

            List<ConvexHullCollider> ColliderA = CharacterLocal.GjkCollider.ConvexHulls;
            List<ConvexHullCollider> ColliderB = OtherMagician.GjkCollider.ConvexHulls;

            DbVector3 PositionB = OtherMagician.Position;
            float YawRadiansB = ToRadians(OtherMagician.Rotation.Yaw);

            bool IntersectsMagician = SolveGjk(ColliderA, PositionA, YawRadiansA, ColliderB, PositionB, YawRadiansB, out GjkResult GjkResultMagician);
            if (IntersectsMagician is false) return false;

            DbVector3 CenterAWorld = GetColliderCenterWorld(CharacterLocal.GjkCollider, PositionA, YawRadiansA);
            DbVector3 CenterBWorld = GetColliderCenterWorld(OtherMagician.GjkCollider, PositionB, YawRadiansB);

            if (EpaSolve(GjkResultMagician, ColliderA, PositionA, YawRadiansA, ColliderB, PositionB, YawRadiansB, out Contact EpaContact))
            {
                DbVector3 ContactNormal = ComputeContactNormal(EpaContact.Normal, CenterAWorld, CenterBWorld);
                float PenetrationDepth = EpaContact.Depth;

                Contacts.Add(new CollisionContact(ContactNormal, PenetrationDepth, CollisionEntryType.Magician));
                return true;
            }
        }

        if (CollisionEntry.Type == CollisionEntryType.Map)
        {
            Map MapPiece = Ctx.Db.Map.Id.Find(CollisionEntry.Id) ?? throw new Exception("Colliding Map Piece Not Found");

            List<ConvexHullCollider> ColliderA = CharacterLocal.GjkCollider.ConvexHulls;
            List<ConvexHullCollider> ColliderB = MapPiece.GjkCollider.ConvexHulls;

            DbVector3 PositionB = new DbVector3(0f, 0f, 0f);
            float YawRadiansB = 0f;

            bool IntersectsMap = SolveGjk(ColliderA, PositionA, YawRadiansA, ColliderB, PositionB, YawRadiansB, out GjkResult GjkResultMap);
            if (IntersectsMap is false) return false;

            DbVector3 CenterAWorld = GetColliderCenterWorld(CharacterLocal.GjkCollider, PositionA, YawRadiansA);
            DbVector3 CenterBWorld = GetColliderCenterWorld(MapPiece.GjkCollider, PositionB, YawRadiansB);

            if (EpaSolve(GjkResultMap, ColliderA, PositionA, YawRadiansA, ColliderB, PositionB, YawRadiansB, out Contact EpaContact))
            {
                DbVector3 ContactNormal = ComputeContactNormal(EpaContact.Normal, CenterAWorld, CenterBWorld);
                float PenetrationDepth = EpaContact.Depth;

                Contacts.Add(new CollisionContact(ContactNormal, PenetrationDepth, CollisionEntryType.Map));
                return true;
            }
        }

        return false;
    }

    static bool TryForceOverlapForEntry(ReducerContext Ctx, ref Magician Character, CollisionEntry Entry, bool WasGrounded)
    {
        if (Entry.Type != CollisionEntryType.Map) return false;
        if (WasGrounded is false && Character.KinematicInformation.Grounded is false) return false;

        float UpwardVelocityBlockThreshold = 0.03f;
        if (Character.Velocity.y > UpwardVelocityBlockThreshold) return false;

        DbVector3 WorldUp = new DbVector3(0f, 1f, 0f);

        float MinGroundDot = 0.75f;
        float FloorUpDot = 0.98f;

        float MaxVerticalGapRamp = 0.045f;
        float MaxVerticalSnap = 0.01f;

        float TinyOverlap = 0.0005f;
        float OverlapEnableGap = 0.01f;

        ComplexCollider ColliderAComplex = Character.GjkCollider;
        List<ConvexHullCollider> ColliderA = ColliderAComplex.ConvexHulls;

        Map MapPiece = Ctx.Db.Map.Id.Find(Entry.Id) ?? throw new Exception("Colliding Map Piece Not Found");
        ComplexCollider ColliderBComplex = MapPiece.GjkCollider;
        List<ConvexHullCollider> ColliderB = ColliderBComplex.ConvexHulls;

        DbVector3 PositionA = Character.Position;
        DbVector3 PositionB = new DbVector3(0f, 0f, 0f);

        float YawA = ToRadians(Character.Rotation.Yaw);
        float YawB = 0f;

        if (SolveGjkDistance(ColliderA, PositionA, YawA, ColliderB, PositionB, YawB, out GjkDistanceResult DistanceResult) is false)
            return false;

        DbVector3 CenterAWorld = GetColliderCenterWorld(ColliderAComplex, PositionA, YawA);
        DbVector3 CenterBWorld = GetColliderCenterWorld(ColliderBComplex, PositionB, YawB);

        DbVector3 ContactNormal = ComputeContactNormal(DistanceResult.SeparationDirection, CenterAWorld, CenterBWorld);

        float UpDot = Dot(ContactNormal, WorldUp);
        if (UpDot < MinGroundDot) return false;

        if (UpDot > FloorUpDot) return false;

        DbVector3 Delta = Sub(DistanceResult.PointOnA, DistanceResult.PointOnB);
        float VerticalGap = Dot(Delta, WorldUp);

        if (VerticalGap <= 0f) return false;
        if (VerticalGap > MaxVerticalGapRamp) return false;

        float SnapDown = VerticalGap;
        if (VerticalGap <= OverlapEnableGap)
            SnapDown = VerticalGap + TinyOverlap;

        if (SnapDown > MaxVerticalSnap)
            SnapDown = MaxVerticalSnap;

        if (SnapDown <= 1e-6f) return false;

        Character.Position = Add(Character.Position, Mul(WorldUp, -SnapDown));
        return true;
    }
}

File: server-rust/spacetimedb/src/Magician/Methods.rs
use std::time::Duration;
use spacetimedb::{rand::Rng, Identity, SpacetimeType, ReducerContext, ScheduleAt, Table, Timestamp};

pub fn AdjustGrounded(ctx: &ReducerContext, move_velocity: &DbVector3, magician: &mut Magician) 
{
    if magician.kinematic_information.grounded {
        magician.kinematic_information.falling = false;
        let can_jump_entry = GetPermissionEntry(&mut magician.player_permission_config, "CanJump").expect("Permission entry not found: CanJump");
        RemoveSubscriber(&mut can_jump_entry.subscribers, "Jump");

        let can_crouch_entry = GetPermissionEntry(&mut magician.player_permission_config, "CanCrouch").expect("Permission entry not found: CanCrouch");
        RemoveSubscriber(&mut can_crouch_entry.subscribers, "Jump");
    } 
    
    else {
        magician.kinematic_information.falling = move_velocity.y < -2.0;
        let can_jump_entry = GetPermissionEntry(&mut magician.player_permission_config, "CanJump").expect("Permission entry not found: CanJump");
        AddSubscriberUnique(&mut can_jump_entry.subscribers, "Jump");

        let can_crouch_entry = GetPermissionEntry(&mut magician.player_permission_config, "CanCrouch").expect("Permission entry not found: CanCrouch");
        AddSubscriberUnique(&mut can_crouch_entry.subscribers, "Jump");
    }
}

pub fn ResolveContacts(magician: &mut Magician, contacts: &Vec<CollisionContact>, input_velocity: &DbVector3) {
    let world_up = DbVector3 { x: 0.0, y: 1.0, z: 0.0 };
    let min_ground_dot: f32 = 0.75;
    let depth_epsilon: f32 = 2e-3;
    let max_depth: f32 = 0.08;
    let correction_factor: f32 = 0.5;
    let target_penetration: f32 = 0.01;
    let max_position_correction: f32 = 0.015;
    let ground_stick_up_threshold: f32 = 0.03;
    let input_up_cancel_threshold: f32 = 0.03;

    let mut corrected_velocity = input_velocity.clone();
    let mut is_grounded_on_map: bool = false;
    let mut ground_support_normal = world_up.clone();
    let mut has_any_position_correction: bool = false;
    let mut total_position_correction = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };

    for contact in contacts.iter() {
        let normal = contact.normal;

        let up_dot: f32 = Dot(&normal, &world_up);
        let is_walkable_map_contact: bool = contact.collision_type == CollisionEntryType::Map && up_dot >= min_ground_dot;

        if is_walkable_map_contact {
            is_grounded_on_map = true;
            ground_support_normal = normal;
        }

        let normal_velocity_component: f32 = Dot(&normal, &corrected_velocity);
        if normal_velocity_component < 0.0 {
            corrected_velocity = Sub(&corrected_velocity, &Mul(&normal, normal_velocity_component));
        }

        let mut depth: f32 = contact.penetration_depth - target_penetration;
        if depth > depth_epsilon {
            if depth > max_depth { depth = max_depth; }
            has_any_position_correction = true;
            total_position_correction = Add(&total_position_correction, &Mul(&normal, depth));
        }
    }

    if has_any_position_correction {
        let correction_magnitude_sq: f32 = Dot(&total_position_correction, &total_position_correction);
        if correction_magnitude_sq > 1e-8 {
            let mut total_position_correction_local = total_position_correction.clone();
            let correction_magnitude: f32 = correction_magnitude_sq.sqrt();

            if correction_magnitude > max_position_correction {
                total_position_correction_local = Mul(&Normalize(&total_position_correction_local), max_position_correction);
            }

            total_position_correction_local = Mul(&total_position_correction_local, correction_factor);
            magician.position = Add(&magician.position, &total_position_correction_local);
        }
    }

    if is_grounded_on_map {
        let desired_horizontal_speed_sq: f32 = input_velocity.x * input_velocity.x + input_velocity.z * input_velocity.z;

        if desired_horizontal_speed_sq < 0.001 {
            corrected_velocity.x = 0.0;
            corrected_velocity.z = 0.0;
        } else {
            let desired_horizontal_speed: f32 = desired_horizontal_speed_sq.sqrt();
            let preserved_corrected_y: f32 = corrected_velocity.y;

            let tangent_velocity = Sub(&corrected_velocity, &Mul(&ground_support_normal, Dot(&corrected_velocity, &ground_support_normal)));
            let tangent_speed_sq: f32 = Dot(&tangent_velocity, &tangent_velocity);

            if tangent_speed_sq > 1e-8 {
                let tangent_direction = Mul(&tangent_velocity, 1.0 / tangent_speed_sq.sqrt());

                let tangent_direction_horizontal_sq: f32 = tangent_direction.x * tangent_direction.x + tangent_direction.z * tangent_direction.z;
                if tangent_direction_horizontal_sq > 1e-8 {
                    let tangent_direction_horizontal: f32 = tangent_direction_horizontal_sq.sqrt();
                    let scale: f32 = desired_horizontal_speed / tangent_direction_horizontal;

                    let desired_tangent_velocity = Mul(&tangent_direction, scale);

                    corrected_velocity.x = desired_tangent_velocity.x;
                    corrected_velocity.z = desired_tangent_velocity.z;
                    corrected_velocity.y = preserved_corrected_y;
                }
            }
        }

        if corrected_velocity.y <= ground_stick_up_threshold { corrected_velocity.y = 0.0; }
        if input_velocity.y <= input_up_cancel_threshold { magician.velocity.y = 0.0; }
    }

    magician.is_colliding = contacts.len() > 0;
    magician.corrected_velocity = corrected_velocity;
    magician.kinematic_information.grounded = magician.kinematic_information.grounded || is_grounded_on_map;
}

pub fn TryBuildContactForEntry(ctx: &ReducerContext, character_local: &Magician, collision_entry: &CollisionEntry, contacts: &mut Vec<CollisionContact>) -> bool {
    let position_a = character_local.position;
    let yaw_radians_a: f32 = ToRadians(character_local.rotation.yaw);

    if collision_entry.entry_type == CollisionEntryType::Magician {
        let other_magician = ctx.db.magician().id().find(collision_entry.id).expect("Colliding Magician Not Found");
        if other_magician.id == character_local.id { return false; }

        let collider_a: &Vec<ConvexHullCollider> = &character_local.collider.convex_hulls;
        let collider_b: &Vec<ConvexHullCollider> = &other_magician.collider.convex_hulls;

        let position_b = other_magician.position;
        let yaw_radians_b: f32 = ToRadians(other_magician.rotation.yaw);

        let mut gjk_result_magician: GjkResult = Default::default();
        let intersects_magician: bool = SolveGjk(collider_a, position_a, yaw_radians_a, collider_b, position_b, yaw_radians_b, &mut gjk_result_magician);
        if intersects_magician == false { return false; }

        let center_a_world = GetColliderCenterWorld(&character_local.collider, position_a, yaw_radians_a);
        let center_b_world = GetColliderCenterWorld(&other_magician.collider, position_b, yaw_radians_b);

        let mut epa_contact: Contact = Default::default();
        if EpaSolve(&gjk_result_magician, collider_a, position_a, yaw_radians_a, collider_b, position_b, yaw_radians_b, &mut epa_contact) {
            let contact_normal = ComputeContactNormal(epa_contact.normal, center_a_world, center_b_world);
            let penetration_depth = epa_contact.depth;
            contacts.push(CollisionContact { normal: contact_normal, penetration_depth, collision_type: CollisionEntryType::Magician });
            return true;
        }
    }

    if collision_entry.entry_type == CollisionEntryType::Map {
        let map_piece = ctx.db.map().id().find(collision_entry.id).expect("Colliding Map Piece Not Found");

        let collider_a: &Vec<ConvexHullCollider> = &character_local.collider.convex_hulls;
        let collider_b: &Vec<ConvexHullCollider> = &map_piece.collider.convex_hulls;

        let position_b = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };
        let yaw_radians_b: f32 = 0.0;

        let mut gjk_result_map: GjkResult = Default::default();
        let intersects_map: bool = SolveGjk(collider_a, position_a, yaw_radians_a, collider_b, position_b, yaw_radians_b, &mut gjk_result_map);
        if intersects_map == false { return false; }

        let center_a_world = GetColliderCenterWorld(&character_local.collider, position_a, yaw_radians_a);
        let center_b_world = GetColliderCenterWorld(&map_piece.collider, position_b, yaw_radians_b);

        let mut epa_contact: Contact = Default::default();
        if EpaSolve(&gjk_result_map, collider_a, position_a, yaw_radians_a, collider_b, position_b, yaw_radians_b, &mut epa_contact) {
            let contact_normal = ComputeContactNormal(epa_contact.normal, center_a_world, center_b_world);
            let penetration_depth = epa_contact.depth;
            contacts.push(CollisionContact { normal: contact_normal, penetration_depth, collision_type: CollisionEntryType::Map });
            return true;
        }
    }

    false
}

pub fn TryForceOverlapForEntry(ctx: &ReducerContext, character: &mut Magician, entry: &CollisionEntry, was_grounded: bool) -> bool {
    if entry.entry_type != CollisionEntryType::Map { return false; }
    if was_grounded == false && character.kinematic_information.grounded == false { return false; }

    let upward_velocity_block_threshold: f32 = 0.03;
    if character.velocity.y > upward_velocity_block_threshold { return false; }

    let world_up = DbVector3 { x: 0.0, y: 1.0, z: 0.0 };

    let min_ground_dot: f32 = 0.75;
    let floor_up_dot: f32 = 0.98;

    let max_vertical_gap_ramp: f32 = 0.045;
    let max_vertical_snap: f32 = 0.01;

    let tiny_overlap: f32 = 0.0005;
    let overlap_enable_gap: f32 = 0.01;

    let collider_a_complex = character.collider;
    let collider_a: &Vec<ConvexHullCollider> = &collider_a_complex.convex_hulls;

    let map_piece = ctx.db.map().id().find(entry.id).expect("Colliding Map Piece Not Found");
    let collider_b_complex = map_piece.collider;
    let collider_b: &Vec<ConvexHullCollider> = &collider_b_complex.convex_hulls;

    let position_a = character.position;
    let position_b = DbVector3 { x: 0.0, y: 0.0, z: 0.0 };

    let yaw_a: f32 = ToRadians(character.rotation.yaw);
    let yaw_b: f32 = 0.0;

    let mut distance_result: GjkDistanceResult = Default::default();
    if SolveGjkDistance(collider_a, position_a, yaw_a, collider_b, position_b, yaw_b, &mut distance_result) == false { return false; }

    let center_a_world = GetColliderCenterWorld(&collider_a_complex, position_a, yaw_a);
    let center_b_world = GetColliderCenterWorld(&collider_b_complex, position_b, yaw_b);

    let contact_normal = ComputeContactNormal(distance_result.separation_direction, center_a_world, center_b_world);

    let up_dot: f32 = Dot(&contact_normal, &world_up);
    if up_dot < min_ground_dot { return false; }
    if up_dot > floor_up_dot { return false; }

    let delta = Sub(&distance_result.point_on_a, &distance_result.point_on_b);
    let vertical_gap: f32 = Dot(&delta, &world_up);

    if vertical_gap <= 0.0 { return false; }
    if vertical_gap > max_vertical_gap_ramp { return false; }

    let mut snap_down: f32 = vertical_gap;
    if vertical_gap <= overlap_enable_gap { snap_down = vertical_gap + tiny_overlap; }
    if snap_down > max_vertical_snap { snap_down = max_vertical_snap; }
    if snap_down <= 1e-6 { return false; }

    character.position = Add(&character.position, &Mul(&world_up, -snap_down));
    true
}

pub fn TryReload(ctx: &ReducerContext, magician: &mut Magician) {
    let bullet_capacity: i32 = magician.bullet_capacity;
    let missing_bullets: i32 = bullet_capacity - magician.bullets.len() as i32;
    if missing_bullets <= 0 { return; }

    let mut new_bullets: Vec<ThrowingCard> = Vec::with_capacity(missing_bullets as usize);
    for _bullet_index in 0..missing_bullets {
        new_bullets.push(CreateThrowingCard());
    }

    magician.bullets.splice(0..0, new_bullets);
}

pub fn TryPerformAttack(ctx: &ReducerContext, magician: &mut Magician, attack_information: AttackInformation) {
    let last_index: usize = magician.bullets.len() - 1;
    let bullet: ThrowingCard = magician.bullets[last_index].clone();
    magician.bullets.remove(last_index);

    let magician_position = magician.position;

    let magician_yaw_radians: f32 = ToRadians(magician.rotation.yaw);
    let magician_yaw_only = Quat::from_rotation_y(magician_yaw_radians);

    let spawn_point = Add(magician_position, Rotate(attack_information.spawn_point_offset, magician_yaw_only));

    let camera_yaw_radians: f32 = ToRadians(magician.rotation.yaw + attack_information.camera_yaw_offset);
    let camera_pitch_radians: f32 = ToRadians(magician.rotation.pitch + attack_information.camera_pitch_offset);
    let camera_rotation = Quat::from_euler(glam::EulerRot::YXZ, camera_yaw_radians, camera_pitch_radians, 0.0);

    let camera_position = Add(magician_position, Rotate(attack_information.camera_position_offset, camera_rotation));
    let camera_forward = Normalize(Rotate(DbVector3 { x: 0.0, y: 0.0, z: 1.0 }, camera_rotation));

    let camera_hit = RaycastMatch(ctx, camera_position, camera_forward, attack_information.max_distance);
    let aim_point = if camera_hit.hit { camera_hit.hit_point } else { Add(camera_position, Mul(camera_forward, attack_information.max_distance)) };

    let shot_delta = Sub(aim_point, spawn_point);
    let shot_direction = Normalize(shot_delta);
    let shot_hit = RaycastMatch(ctx, spawn_point, shot_direction, attack_information.max_distance);

    if shot_hit.hit {
        log::info!("Hitscan Hit Type={} Distance={} EntityId={}", shot_hit.hit_type, shot_hit.hit_distance, shot_hit.hit_entity_id);
    } 
    
    else {
        log::info!("Hitscan Miss");
    }
}

pub fn ResetAllTimers(magician: &mut Magician) {
    for timer_index in 0..magician.timers.len() {
        let mut current_timer = magician.timers[timer_index].clone();
        current_timer.current_time = current_timer.reset_time;
        magician.timers[timer_index] = current_timer;
    }
}

pub fn TryFindTimerIndex(magician: &Magician, timer_name: &str) -> usize {
    for timer_index in 0..magician.timers.len() {
        if magician.timers[timer_index].name == timer_name { return timer_index; }
    }
    panic!("Requested Timer Not Found");
}